simpleideo = function() structure(list(Chromosome = structure(c(1L, 12L, 16L, 17L, 18L, 
19L, 20L, 21L, 22L, 23L, 24L, 2L, 3L, 4L, 5L, 6L, 7L, 8L, 9L, 
10L, 11L, 13L, 14L, 15L), .Label = c("chr1", "chr10", "chr11", 
"chr12", "chr13", "chr14", "chr15", "chr16", "chr17", "chr18", 
"chr19", "chr2", "chr20", "chr21", "chr22", "chr3", "chr4", "chr5", 
"chr6", "chr7", "chr8", "chr9", "chrX", "chrY"), class = "factor"), 
    ChromStart = c(0L, 0L, 0L, 0L, 0L, 0L, 0L, 0L, 0L, 0L, 0L, 
    0L, 0L, 0L, 0L, 0L, 0L, 0L, 0L, 0L, 0L, 0L, 0L, 0L), ChromEnd = c(249250621L, 
    243199373L, 198022430L, 191154276L, 180915260L, 171115067L, 
    159138663L, 146364022L, 141213431L, 155270560L, 59373566L, 
    135534747L, 135006516L, 133851895L, 115169878L, 107349540L, 
    102531392L, 90354753L, 81195210L, 78077248L, 59128983L, 63025520L, 
    48129895L, 51304566L), Band = structure(c(97L, 72L, 75L, 
    45L, 40L, 72L, 59L, 66L, 69L, 62L, 11L, 37L, 42L, 29L, 20L, 
    20L, 20L, 26L, 26L, 11L, 26L, 20L, 20L, 20L), .Label = c("p11", 
    "p11.1", "p11.11", "p11.12", "p11.2", "p11.21", "p11.22", 
    "p11.23", "p11.3", "p11.31", "p11.32", "p11.4", "p12", "p12.1", 
    "p12.2", "p12.3", "p12.31", "p12.32", "p12.33", "p13", "p13.1", 
    "p13.11", "p13.12", "p13.13", "p13.2", "p13.3", "p13.31", 
    "p13.32", "p13.33", "p14", "p14.1", "p14.2", "p14.3", "p15", 
    "p15.1", "p15.2", "p15.3", "p15.31", "p15.32", "p15.33", 
    "p15.4", "p15.5", "p16.1", "p16.2", "p16.3", "p21", "p21.1", 
    "p21.2", "p21.3", "p21.31", "p21.32", "p21.33", "p22", "p22.1", 
    "p22.11", "p22.12", "p22.13", "p22.2", "p22.3", "p22.31", 
    "p22.32", "p22.33", "p23", "p23.1", "p23.2", "p23.3", "p24.1", 
    "p24.2", "p24.3", "p25.1", "p25.2", "p25.3", "p26.1", "p26.2", 
    "p26.3", "p31.1", "p31.2", "p31.3", "p32.1", "p32.2", "p32.3", 
    "p33", "p34.1", "p34.2", "p34.3", "p35.1", "p35.2", "p35.3", 
    "p36.11", "p36.12", "p36.13", "p36.21", "p36.22", "p36.23", 
    "p36.31", "p36.32", "p36.33", "q11", "q11.1", "q11.2", "q11.21", 
    "q11.22", "q11.221", "q11.222", "q11.223", "q11.23", "q12", 
    "q12.1", "q12.11", "q12.12", "q12.13", "q12.2", "q12.3", 
    "q13", "q13.1", "q13.11", "q13.12", "q13.13", "q13.2", "q13.3", 
    "q13.31", "q13.32", "q13.33", "q13.4", "q13.41", "q13.42", 
    "q13.43", "q13.5", "q14", "q14.1", "q14.11", "q14.12", "q14.13", 
    "q14.2", "q14.3", "q15", "q15.1", "q15.2", "q15.3", "q16.1", 
    "q16.2", "q16.3", "q21", "q21.1", "q21.11", "q21.12", "q21.13", 
    "q21.2", "q21.21", "q21.22", "q21.23", "q21.3", "q21.31", 
    "q21.32", "q21.33", "q22", "q22.1", "q22.11", "q22.12", "q22.13", 
    "q22.2", "q22.3", "q22.31", "q22.32", "q22.33", "q23", "q23.1", 
    "q23.2", "q23.3", "q23.31", "q23.32", "q23.33", "q24", "q24.1", 
    "q24.11", "q24.12", "q24.13", "q24.2", "q24.21", "q24.22", 
    "q24.23", "q24.3", "q24.31", "q24.32", "q24.33", "q25", "q25.1", 
    "q25.2", "q25.3", "q25.31", "q25.32", "q25.33", "q26", "q26.1", 
    "q26.11", "q26.12", "q26.13", "q26.2", "q26.3", "q26.31", 
    "q26.32", "q26.33", "q27", "q27.1", "q27.2", "q27.3", "q28", 
    "q28.1", "q28.2", "q28.3", "q29", "q31.1", "q31.2", "q31.21", 
    "q31.22", "q31.23", "q31.3", "q31.31", "q31.32", "q31.33", 
    "q32", "q32.1", "q32.11", "q32.12", "q32.13", "q32.2", "q32.3", 
    "q32.31", "q32.32", "q32.33", "q33", "q33.1", "q33.2", "q33.3", 
    "q34", "q34.1", "q34.11", "q34.12", "q34.13", "q34.2", "q34.3", 
    "q35", "q35.1", "q35.2", "q35.3", "q36.1", "q36.2", "q36.3", 
    "q37.1", "q37.2", "q37.3", "q41", "q42.11", "q42.12", "q42.13", 
    "q42.2", "q42.3", "q43", "q44"), class = "factor"), Stain = structure(c(2L, 
    2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 
    2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L), .Label = c("acen", "gneg", 
    "gpos100", "gpos25", "gpos50", "gpos75", "gvar", "stalk"), class = "factor")), .Names = c("Chromosome", 
"ChromStart", "ChromEnd", "Band", "Stain"), row.names = c("chr1", 
"chr2", "chr3", "chr4", "chr5", "chr6", "chr7", "chr8", "chr9", 
"chrX", "chrY", "chr10", "chr11", "chr12", "chr13", "chr14", 
"chr15", "chr16", "chr17", "chr18", "chr19", "chr20", "chr21", 
"chr22"), class = "data.frame")
 

sglToCircos = function(sgl, chr.exclude=NULL, snpgr=NULL, pltextsize=.6,
    trks.inside=7, linktrknum=4, genetrknum=4, snptrknum=1  ) {
require(bceQTL)
require(SNPlocs.Hsapiens.dbSNP144.GRCh37)
require(Homo.sapiens)

# Given a manually constructed SNP->GENE list, construct circos plot

sglDF = sglistToDF(sgl, snpgr)
myd = data.frame(sglDF[,2], sglDF[,3], sglDF[,3]+1, sglDF[,4], sglDF[,5], sglDF[,5]+1)


library(RCircos)
data(UCSC.HG19.Human.CytoBandIdeogram);
cyto.info <- UCSC.HG19.Human.CytoBandIdeogram;  # from 4
cyto.info = simpleideo()
RCircos.Set.Core.Components(cyto.info, chr.exclude=chr.exclude, tracks.inside=trks.inside,
   tracks.outside=0)
RCircos.Set.Plot.Area();
RCircos.Chromosome.Ideogram.Plot();
RCircos.Link.Plot(myd, track.num=linktrknum)
RCircos.Env$RCircos.PlotPar$text.size = pltextsize

if (any(dd <- duplicated(sglDF$snpid))) {
   normlocs = sglDF[-which(dd),1:3]
   }
else normlocs = sglDF[,1:3]

snplab = data.frame(normlocs[,2], normlocs[,3], normlocs[,3], normlocs[,1], stringsAsFactors=FALSE)
snplab[,1] = as.character(snplab[,1])
snplab[,4] = as.character(snplab[,4])

RCircos.Gene.Name.Plot(snplab,name.col=4,track.num=snptrknum,"in") # from 2

#geneLabelDF = function(syms) {
# require(Homo.sapiens)
# chrs = mapIds(Homo.sapiens, keys=syms, keytype="SYMBOL", column="TXCHROM")
# locs = mapIds(Homo.sapiens, keys=syms, keytype="SYMBOL", column="TXSTART")
# ans = data.frame(chrs, as.numeric(locs), as.numeric(locs), syms)
# ans = na.omit(ans)
# colnames(ans) = c("chr", "start", "start2", "sym")
# ans
#}

genes = na.omit(unlist(sgl))
gdf = geneLabelDF(genes)
RCircos.Gene.Name.Plot(gene.data=gdf,name.col=4,track.num=genetrknum,side="in")
invisible(list(gdf=gdf, snplab=snplab,linkdf=myd))
}
